### ISSUE TITLE
Setting empty object causes validation to fail

### ISSUE THREAD

**Issue by paglias** [Automattic/mongoose#4182]
I have a schema like this:

```
{
  // other fields
  challenge: {
    id: {type: String, ref: 'Challenge', validate: [validator.isUUID, 'Invalid uuid.']}, // When set (and userId not set) it's the original task
    taskId: {type: String, ref: 'Task', validate: [validator.isUUID, 'Invalid uuid.']}, // When not set but challenge.id defined it's the original task
    broken: {type: String, enum: ['CHALLENGE_DELETED', 'TASK_DELETED', 'UNSUBSCRIBED', 'CHALLENGE_CLOSED', 'CHALLENGE_TASK_NOT_FOUND']}, // CHALLENGE_TASK_NOT_FOUND comes from v3 migration
    winner: String, // user.profile.name of the winner
  },
}
```

If doc.challenge has some values defined and then i try to set it to `{}`

```
doc.challenge = {}
doc.save().catch(function (err) {
  console.log(err)
})
```

saving fails with

```
name=ValidationError, 
message=`null` is not a valid enum value for path `challenge.broken`., 
name=ValidatorError, enumValues=[CHALLENGE_DELETED, TASK_DELETED, UNSUBSCRIBED, CHALLENGE_CLOSED, CHALLENGE_TASK_NOT_FOUND], 
type=enum, 
message=`{VALUE}` is not a valid enum value for path `{PATH}`., 
validator=function (v) {return undefined === v || ~vals.indexOf(v);`
```

with similar messages for the other fields that have some type of validation.

The problem seems to be (from `validator=function (v) {return undefined === v || ~vals.indexOf(v);`) that mongoose think `challenge.broken` is null while instead is undefined and then tries to validate the field which of course fails

**Comment by paglias** [comment]
If it can help I'm using `minimize: false` so maybe this is linked https://github.com/Automattic/mongoose/issues/2930

**Comment by paglias** [comment]
@vkarpov15 thanks!

**Comment by paglias** [comment]
@vkarpov15 this is still happening to us with Mongoose 4.4.20

<img width="879" alt="schermata 2016-06-12 alle 12 43 45" src="https://cloud.githubusercontent.com/assets/589911/15990688/62da177a-309b-11e6-806e-705947e33623.png">

With the same sub-schema as before

```
  challenge: {
    id: {type: String, ref: 'Challenge', validate: [validator.isUUID, 'Invalid uuid.']}, 
    taskId: {type: String, ref: 'Task', validate: [validator.isUUID, 'Invalid uuid.']},
    broken: {type: String, enum: ['CHALLENGE_DELETED', 'TASK_DELETED', 'UNSUBSCRIBED', 'CHALLENGE_CLOSED', 'CHALLENGE_TASK_NOT_FOUND']},
    winner: String,
  },
```

and this code (inside a promise)

```
        ....
        task.challenge = {};
        return task.save();
        .....
```

**Comment by vkarpov15** [comment]
This looks like it was fixed with #4218, the below script works just fine for me with master.

```javascript
'use strict';

var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost/gh4182');
mongoose.set('debug', true);

var testSchema = new mongoose.Schema({
  // other fields
  challenge: {
    broken: {type: String, enum: ['CHALLENGE_DELETED', 'TASK_DELETED', 'UNSUBSCRIBED', 'CHALLENGE_CLOSED', 'CHALLENGE_TASK_NOT_FOUND']}, // CHALLENGE_TASK_NOT_FOUND comes from v3 migration
  }
});

var Model = mongoose.model('gh4182', testSchema);

var doc = new Model({ challenge: { broken: 'TASK_DELETED' } });
doc.save(function(error, doc) {
  assert.ifError(error);
  Model.findById(doc, function(error, doc) {
    assert.ifError(error);
    doc.challenge = {};
    doc.save(function(error) {
      assert.ifError(error);
      console.log('done');
      process.exit(0);
    });
  });
});
```

**Comment by paglias** [comment]
Can confirm that this is fixed!
